### PR TITLE
feat(release-activity): backend support for tracking release activity

### PR DESCRIPTION
**Pull request by barkbarkimashark**
Alpha Workflow 2.0 feature for tracking activity when an issue occurs in an active release-deployment time-window. 

This adds the backend components needed to support surfacing the notification activity in the frontend.

### CODE REVIEW

File: src/sentry/notifications/notifications/rules.py
Comment (wedamija) [inline src/sentry/notifications/notifications/rules.py:182]: How often will this get called? `count_users_seen` makes a snuba query, so just want to verify.

File: src/sentry/api/endpoints/project_release_activity.py
Comment (wedamija) [inline src/sentry/api/endpoints/project_release_activity.py:28]: Should we paginate here?

File: src/sentry/models/releaseactivity.py
Comment (wedamija) [inline src/sentry/models/releaseactivity.py:12]: What will we use this type for?

File: src/sentry/models/releaseactivity.py
Comment (barkbarkimashark) [inline src/sentry/models/releaseactivity.py:9]: Tried seeing what this attr is for but to no avail? What is this flag for?

Comment (wedamija) [reply src/sentry/models/releaseactivity.py:9]: It's used for exports here https://github.com/getsentry/sentry/blob/531e83e763186d14c8202190c6fee454e5d0edc4/src/sentry/runner/commands/backup.py#L138. Usually you can leave this as False.

File: src/sentry/models/releaseactivity.py
Comment (wedamija) [reply src/sentry/models/releaseactivity.py:12]: Do we need a default? I think it'd be better to fail the write if a `type` isn't provided

File: src/sentry/api/endpoints/project_release_activity.py
Comment (barkbarkimashark) [reply src/sentry/api/endpoints/project_release_activity.py:17]: Hmm, I tried this but the response returned is slightly different than a regular old call to an non-existent endpoint:
https://github.com/getsentry/sentry/blob/8df0b711532de3a51fd4f3ee0c42a1a90d46a19a/tests/sentry/api/endpoints/test_project_release_activity.py#L12

Basically, if the feature isn't turned on I wanna 'simulate' the exact same response for when the FE calls a non-existent endpoint. 

Using `raise rest_framework.exceptions.NotFound` seems to return `{"detail":"Not found."}` in the response body. 

I'm not sure it matters from the FE perspective whether we return anything in the response body in the case of a 404. What do you think?

File: src/sentry/notifications/notifications/rules.py
Comment (barkbarkimashark) [reply src/sentry/notifications/notifications/rules.py:182]: This path is feature-flagged to our org only and only when a new issue or regression occurs in an active release window. There should be around a 2:1 with the number of times the `ActiveReleaseAlertNotification` triggers to the number of times `group.count_users_seen()` gets called.

File: src/sentry/notifications/notifications/rules.py
Comment (scttcper) [inline src/sentry/notifications/notifications/rules.py:179]: not sure the error_count and issue_affected_users is worth saving

Comment (barkbarkimashark) [reply src/sentry/notifications/notifications/rules.py:179]: If we return the serialized group instead of `issue_id`, `issue_error_count` and `issue_affected_users`, would that still work with the current design?

Comment (scttcper) [reply src/sentry/notifications/notifications/rules.py:179]: i have it as `data.group`, but whatever works

Comment (barkbarkimashark) [reply src/sentry/notifications/notifications/rules.py]: Ok, that works for me. I added a thing.

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (wedamija) [inline src/sentry/api/serializers/models/releaseactivity.py:32]: Mutating the data seems like a risky move, can we just include this in a `group` key or something? If someone happened to save a model after this it'd mess things up.

Alternatively you can just copy data here and use that

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (barkbarkimashark) [inline src/sentry/api/serializers/models/releaseactivity.py:25]: This is really ugly .. 
Instead of serializing the group and storing it in the db, we only store the `group_id` and when we return the response, we'll retrieve the group and serialize it there. 

Not sure if there's a better way.

If we serialize the whole group and persist it in the DB, its essentially snapshoting the group details at the time when it gets stored in the DB vs only storing the `group_id` and retrieving the details with the freshest state.

Comment (wedamija) [inline src/sentry/api/serializers/models/releaseactivity.py:25]: This is still going to make N queries to fetch each group. You want to fetch all the group ids and get them all in one bulk query.

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/releaseactivity.py:32]: Yeah, good call out. I'll make a copy of the data instead of mutating.

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (wedamija) [reply src/sentry/api/serializers/models/releaseactivity.py:25]: Actually just to build on that a bit - you might want to fetch all groups, `serialize` them in one bulk call and then assign them to their respective activity

File: src/sentry/models/deploy.py
Comment (markstory) [inline src/sentry/models/deploy.py:97]: Are you able to use `Organization.objects.get_from_cache()` instead? You should generally use `get()` instead of `first()` as first() [will add an order clause](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#first) which can yield poorer performance than `get()` will.

Given that you are mostly using the organization to check features reading from cache will save us a SQL query.

**Review by AniketDas-Tekky — APPROVED** [review]
Endpoint looks good. Don't merge until someone else approves too.

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (wedamija) [inline src/sentry/api/serializers/models/releaseactivity.py:22]: This would be `MutableMapping[ReleaseActivity, Any]`

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (wedamija) [inline src/sentry/api/serializers/models/releaseactivity.py:43]: Same here for the mapping

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (wedamija) [inline src/sentry/api/serializers/models/releaseactivity.py:35]: If the group has been deleted, should we still return these rows?

**Review by wedamija — APPROVED** [review]

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/releaseactivity.py:35]: Hmmm, didn't think of that. Would it make more sense to put that logic in the FE?

Comment (wedamija) [reply src/sentry/api/serializers/models/releaseactivity.py:35]: If we had a relational link to groups from the table it'd be easy to filter these rows out. Not sure if we need to do that now,  but it might be better to store relational data outside of the json blob if possible

Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/releaseactivity.py:35]: I haven't looked into the semantics of how Group deletions work. Does that mark it as deleted, or does it actually remove the row entirely?

How would one know if a Group is deleted?

Tbh, I hadn't thought enough about the timeline and how much 'state' we should be reflecting there. A group deletion activity might be useful to have in the timeline.

Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/releaseactivity.py:35]: I see there's a GroupStatus here:
https://github.com/getsentry/sentry/blob/94c896a4a3663abbd31775957f1aa5448fde5491/src/sentry/models/group.py#L379-L388

But that doesn't include deletion.

Comment (wedamija) [reply src/sentry/api/serializers/models/releaseactivity.py:35]: Groups are fully removed from the database after deletion, which is either explicit, or after 90 days of no new events.

Actually, I guess if we had a relation to the `Group` table then we'd just automatically have these rows removed once a group is deleted, which also solves the problem (unless you want to keep them... but there's no information about the group, so not sure how useful that is)

Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/releaseactivity.py:35]: Yeah, I dived a bit to see how deletion works and it does look like the rows are completely wiped out:
https://github.com/getsentry/sentry/blob/9ced05b0cf494b8398b5ce1c23957b620a8619ee/src/sentry/deletions/defaults/group.py#L100-L134

In the case where someone deletes the group, the returned data would be `group=None`. I think we could put in some logic in the FE to filter these cases out.

File: src/sentry/api/serializers/models/releaseactivity.py
Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/releaseactivity.py:43]: Hmm, I think the `attrs` is the value of whatever is mapped from `get_attrs`? In this case, itll be the type defined by the `data` 'Mapping[str, JSONData]', right?

https://github.com/getsentry/sentry/blob/9113cb2b7db1b69a7e206d90df93a7404aa6c659/src/sentry/api/serializers/base.py#L117

**Pull request merged**
